src/Utils.ts

```diff
@@ -986,12 +986,129 @@ function setDummyInputDebugValue(
     );
 }
 
+export class DummyInputObserver implements Types.DummyInputObserver {
+    private _win?: GetWindow;
+    private _offsetsQueue: (() => void)[] = [];
+    private _offsetsTimer?: number;
+    private _domChangedMap: WeakMap<HTMLElement, true> = new WeakMap();
+    private _domChangedTimer?: number;
+    private _dummyMap: WeakMap<HTMLElement, () => void> = new WeakMap();
+    private _dummies: WeakRef<HTMLElement>[] = [];
+    domChanged?(parent: HTMLElement | null): void;
+
+    constructor(win: GetWindow) {
+        this._win = win;
+    }
+
+    add(dummy: HTMLElement | undefined, callback: () => void): void {
+        if (dummy) {
+            const dummyMap = this._dummyMap;
+            const cb = dummyMap.get(dummy);
+
+            if (!cb) {
+                this._dummies.push(new WeakRef(dummy));
+                this.domChanged = this._domChanged;
+            }
+
+            dummyMap.set(dummy, callback);
+        }
```

Comment:
```suggestion
        if (!dummy) {
          return
        }
        const dummyMap = this._dummyMap;
        const cb = dummyMap.get(dummy);

        if (!cb) {
            this._dummies.push(new WeakRef(dummy));
            this.domChanged = this._domChanged;
        }

        dummyMap.set(dummy, callback);
```

avoid adding another level of nesting by short returning (the less nested the easier to read)

Reply:
`if (false) { return; } doThings();'` is less bundle-size friendly than `if (true) { doThings(); }`  — that is practically the only reason. But sure.